assets/src/edit-story/app/api/apiProvider.js

```diff
@@ -157,12 +158,22 @@ function APIProvider({ children }) {
    */
   const uploadMedia = useCallback(
     (file, additionalData) => {
+      // Intercept window.fetch to support progressive upload indicator
+      apiFetch.setFetchHandler((options) => {
```

Comment:
Hey WP folks @dvoytenko @spacedmonkey @swissspidy @barklund , since `apiFetch` use under the hood `window.fetch`, which doesn't support us to track properly upload progress to match our Uploading UI, is `apiFetch.setFetchHandler` a good way to go? The initial idea is to use something like `axios` that provide a [good path to support uploading progress](https://github.com/axios/axios/blob/6a4a85c57fcaba912eee61b87ba34d07323bc60c/examples/upload/index.html#L30). 

Currently, I'm able to upload media properly, even getting the following error in the console `The response is not a valid JSON response.` after complete the upload. Not sure what it can be yet, I am still investigating, so feel free to also give some input on it as well.  Thanks!

Reply:
I can't talk about WP specific support, but I think we should just do the "infinite progress" as a first step and file a separate issue for upload progress. We can look into axios or any other approach in that issue.

Reply:
This is what I recommended from my research. How do you imagine tracking in the progress? New state in the reducer?

Reply:
@spacedmonkey I was able to fix the error `The response is not a valid JSON response.` normalizing the HTTP response from axios using a `.json()` mock to simulate it to `apiFetch` internal validations, but after some tests, it started to raise side effects in parts not related to the context of this PR. So for now, I will do as @dvoytenko suggested. Happyfully, the Uploading UI proposal already expects an "infinite loading" behavior. 

From here, I will apply some improvements based on your PR comments guys, solve potential conflicts after merging with master since there are changes and open this PR. Thank you guys.